opensearch_single_kernel/managers/lock.py

```diff
@@ -0,0 +1,507 @@
+                # if the charm code raises an uncaught exception later in the Juju event.
+                self.logger.debug(
+                    "[Node lock] Single unit deployment. Not waiting until next Juju event to use peer "
+                    "databag lock for leader unit"
+                )
+            else:
+                self.logger.debug(
+                    "[Node lock] Not acquired. Waiting until next Juju event to use peer databag lock "
+                    "for leader unit"
+                )
+                return False
+
+        self.logger.debug("[Node lock] Acquired via peer databag")
+        return True
+
+    def release(self):
+        """Release lock for this unit."""
+        if not self._relation:
+            return
+
+        self._relation.data[self._charm.unit].pop("lock-requested", None)
+        if self.state.opensearch_unit.is_app_leader:
+            self.logger.debug("[Node lock] Released peer lock as leader unit")
+            # A separate relation-changed event won't get fired
+            self._on_peer_relation_changed()
+
+    def _unit_requested_lock(self, unit: ops.Unit):
+        """Whether unit requested lock."""
+        assert self._relation
+        if not (value := self._relation.data.get(unit, {}).get("lock-requested")):
+            return False
+
+        value = json.loads(value)
+        if not isinstance(value, bool):
+            raise ValueError
+
+        return value
+
+    @property
+    def _unit_with_lock(self) -> str | None:
+        if self._relation:
+            return self._relation.data[self.state.opensearch_application.app].get("unit-with-lock")
+
+    @_unit_with_lock.setter
+    def _unit_with_lock(self, value: str):
+        assert self._relation
+        assert self._unit_with_lock != value
+
+        if value == self.state.opensearch_unit.unit_name:
+            self.logger.debug("[Node lock] (leader) granted peer lock to own unit")
+            # Prevent leader unit from using lock in the same Juju event that it was granted
+            # If the charm code raises an uncaught exception later in the Juju event,
+            # `unit-with-lock` will be reverted to its previous value—which could allow another
+            # unit to get the lock.
+            # Therefore, we cannot use the lock in this Juju event. We must wait until the next
+            # Juju event, when `unit-with-lock` has been committed (i.e. won't be reverted), to use
+            # the lock.
+            # `JUJU_CONTEXT_ID` is unique for each Juju event
+            # (https://matrix.to/#/!xdClnUGkurzjxqiQcN:ubuntu.com/$yEGjGlDaIPBtCi8uB3fH6ZaXUjN7GF-Y2s9YwvtPM-o?via=ubuntu.com&via=matrix.org&via=cutefunny.art)
+
+            self._relation.data[self.state.opensearch_application.app][
+                "leader-acquired-lock-after-juju-event-id"
+            ] = os.environ["JUJU_CONTEXT_ID"]
+        self._relation.data[self.state.opensearch_application.app]["unit-with-lock"] = value
```

Comment:
```suggestion
 ] = os.environ["JUJU_CONTEXT_ID"] 
self._relation.data[self.state.application.app]["unit-with-lock"] = value
```
if JUJU_CONTEXT_ID env variable is not defined, the value will be None. We need to handle this.

Reply:
This can be addressed in a separate PR since this might be a bug in OpenSearch